sphinx/source/bokeh/static/searchtools.js

```diff
@@ -0,0 +1,534 @@
+        for (j = 0; j < _files.length; j++) {
+          file = _files[j];
+          if (!(file in scoreMap))
+            scoreMap[file] = {};
+          scoreMap[file][word] = o.score;
+        }
+      });
+
+      // create the mapping
+      for (j = 0; j < files.length; j++) {
+        file = files[j];
+        if (file in fileMap && fileMap[file].indexOf(word) === -1)
+          fileMap[file].push(word);
+        else
+          fileMap[file] = [word];
+      }
+    }
+
+    // now check if the files don't contain excluded terms
+    for (file in fileMap) {
+      var valid = true;
+
+      // check if all requirements are matched
+      var filteredTermCount = // as search terms with length < 3 are discarded: ignore
+        searchterms.filter(function(term){return term.length > 2}).length
+      if (
+        fileMap[file].length != searchterms.length &&
+        fileMap[file].length != filteredTermCount
+      ) continue;
+
+      // ensure that none of the excluded terms is in the search result
+      for (i = 0; i < excluded.length; i++) {
+        if (terms[excluded[i]] == file ||
+            titleterms[excluded[i]] == file ||
+            $u.contains(terms[excluded[i]] || [], file) ||
+            $u.contains(titleterms[excluded[i]] || [], file)) {
+          valid = false;
+          break;
+        }
+      }
+
+      // if we have still a valid result we can add it to the result list
+      if (valid) {
+        // select one (max) score for the file.
+        // for better ranking, we should calculate ranking by using words statistics like basic tf-idf...
+        var score = $u.max($u.map(fileMap[file], function(w){return scoreMap[file][w]}));
+        results.push([docnames[file], titles[file], '', null, score, filenames[file]]);
+      }
+    }
+    return results;
+  },
+
+  /**
+   * helper function to return a node containing the
+   * search summary for a given text. keywords is a list
+   * of stemmed words, hlwords is the list of normal, unstemmed
+   * words. the first one is used to find the occurrence, the
+   * latter for highlighting it.
+   */
+  makeSearchSummary : function(htmlText, keywords, hlwords) {
+    var text = Search.htmlToText(htmlText);
+    var textLower = text.toLowerCase();
+    var start = 0;
+    $.each(keywords, function() {
+      var i = textLower.indexOf(this.toLowerCase());
+      if (i > -1)
+        start = i;
+    });
+    start = Math.max(start - 120, 0);
+    var excerpt = ((start > 0) ? '...' : '') +
+      $.trim(text.substr(start, 240)) +
+      ((start + 240 - text.length) ? '...' : '');
+    var rv = $('<div class="context"></div>').text(excerpt);
+    $.each(hlwords, function() {
+      rv = rv.highlightText(this, 'highlighted');
+    });
+    return rv;
+  }
+};
+
+$(document).ready(function() {
+  var para = document.getElementsByTagName('p')[1]
+  para.innerHTML+=`<br><input type="checkbox" id="concise">&nbsp;&nbsp;Check this option for faster, but shorter querying<br>`
```

Comment:
Would it be simple to just automatically show a page with "fast" results and then have a button to go to a page with "full" results? I think sometimes less choice is better, and I worry that this checkbox will be unfamiliar and confusing (or ignored) by users

Reply:
@bryevdv Hmm. Maybe. 

Introducing a button onto a page (which fills up quickly with search queries) I felt was a bad idea because if I position the button to appear after the search results, it makes no sense. The user will have to scroll to the end anyways. 

If I position the button next to the actual search button, users are bound to get confused which one to click on, and the UI doesn't look good when you have 2 buttons placed next to each other doing pretty much the same function. That's why I went with a checkbox, because that's the simplest way to get me a boolean value for my purposes. 

I hope the checkbox is not too badly placed/ ill-purposed. I tried multiple things to fit in better, but this was the one I liked the most.

Reply:
My main objection to the checkbox is that I can't think of any search context I've ever seen that has anything like that. It's very novel, and I think in this context novel is not desirable. If the fast results are limited to only show a smaller number (say less than 30) then I think scrolling to the bottom is not much of an issue? It seems very familiar/similar to me to e.g. scrolling to the bottom of Google results to get the to next page.

Reply:
I should add: in my experience  the result I want when searching is usually in the top 10 or even 5 results, so I think limiting a "first page" to ~30 results would be more than enough to cover the vast majority of needs (and still make the "next page" button reasonable to get to for anyone that needs more)

Reply:
@bryevdv You're right, in the sense that a checkbox isn't the most user friendly way to go about it. I wish I could do better than it, but I do believe it fixes the basic issue of slower searches, a slightly better UX. 

I think ultimately the bokeh docs search calls for good "pagination". At this time, what I've done is I've spliced the results and I'm showing the top half of them (for the fast query option). In case I've to show the top 30 results, I'd need to find a way to either:-

1. Hide the remaining results using some CSS and then upon a button click, reveal them.
2. Store the results somewhere and later append them into the page if the user wants more results to be shown.
3. Show the results page by page.

As of now, I spent some time getting familiar with sphinx and bokeh, and therefore made an attempt at improving the UX/Docs experince by a little bit. Yes, I did try the CSS approach too, but I couldn't get it to work. There must be some clever/smart way of doing it. I tried `nth-child` method so that all query results above 30 would instantly be hidden. It didn't work as per expectation. I should definitely look into how popular search engines do this kind of work so efficienty!

Option 3 would be absolutely great. Given time, I think its definitely worth a go. Maybe this could be  GSoD project idea in itself, coupled with something like "Accesibility a11y improvement". 

Thanks!